core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -478,34 +479,31 @@ private[log] class Cleaner(val id: Int,
                              activeProducers: Map[Long, ProducerIdEntry],
                              stats: CleanerStats) {
     val logCleanerFilter = new RecordFilter {
-      var retainLastBatchSequence: Boolean = false
       var discardBatchRecords: Boolean = false
 
-      override def shouldDiscard(batch: RecordBatch): Boolean = {
+      override def checkBatchRetention(batch: RecordBatch): BatchRetention = {
         // we piggy-back on the tombstone retention logic to delay deletion of transaction markers.
         // note that we will never delete a marker until all the records from that transaction are removed.
         discardBatchRecords = shouldDiscardBatch(batch, transactionMetadata, retainTxnMarkers = retainDeletes)
 
         // check if the batch contains the last sequence number for the producer. if so, we cannot
         // remove the batch just yet or the producer may see an out of sequence error.
         if (batch.hasProducerId && activeProducers.get(batch.producerId).exists(_.lastSeq == batch.lastSequence)) {
-          retainLastBatchSequence = true
-          false
+          BatchRetention.RETAIN_EMPTY
```

Comment:
Now that the cleaner logic is a bit more complicated with respect to transactional data, could we add a bit more documentation on how the new cleaning logic works? For example, it would be useful to document how we clean an aborted batch not on the last sequence, an aborted batch on the last sequence,  a committed batch not on the last sequence, a committed batch on the last sequence,  transaction marker, etc.

Also, it seems that the new logic allows an AbortedMaker to be deleted before the place-holder (for last sequence) aborted batch in the transaction is deleted from the log. This doesn't seem to cause any harm except that it doesn't seem to be consistent with the CommitMarker. It seems that the CommitMaker can't be deleted before the place-holder committed batch is deleted.